feat(nix): run the GC more often and free up more

steveeJ opened the pull request:
we've been running running into disk space issue on CI. this is an attempt to resolve that.

DavHau commented on modules/nixos/shared.nix:
```diff
@@ -29,10 +29,10 @@
     }
-    // lib.optionalAttrs pkgs.stdenv.isLinux (lib.mkForce {dates = "*:30";})
-    // lib.optionalAttrs pkgs.stdenv.isDarwin {interval.Minute = 30;};
+    // lib.optionalAttrs pkgs.stdenv.isLinux (lib.mkForce {dates = "*:17";})
```
That will still run it only once per hour, or am I misunderstanding the format?

steveej replied:
good question. i'm somehow interpreting it as a frequency but that might not be true.

DavHau replied:
I interpret it as a wildcard, so `*:*5` should run it every 10 minutes at minute 5, 15, 25 etc.

steveej replied:
just to point out the obvious but we only use one `*` i.e. `*:17`

steveej commented on modules/nixos/shared.nix:
```diff
@@ -29,10 +29,10 @@
-    // lib.optionalAttrs pkgs.stdenv.isLinux (lib.mkForce {dates = "*:30";})
-    // lib.optionalAttrs pkgs.stdenv.isDarwin {interval.Minute = 30;};
+    // lib.optionalAttrs pkgs.stdenv.isLinux (lib.mkForce {dates = "*:17";})
+    // lib.optionalAttrs pkgs.stdenv.isDarwin {interval.Minute = 17;};
```
i tried to look at the logs and see when GC triggers but couldn't find any occurrence of it under `/var/log`. any idea where this is logged on macos?

DavHau replied:
`log show | grep nix-gc` shows that it is running every 17 minutes

DavHau replied:
Oops, I meant it is running every hour on minute 17, like expected

steveej replied:
(we're commenting on the macos line)

trying this on macos-02:

```
administrator@Administrators-Mac-mini> log show | grep nix-gc                                                               ~
log: too many arguments
```

DavHau commented:
I fixed the settings for linux and added some comment regarding macos via 8baf18b25d76ac8762b2c7459166677fbe8436a6

steveej commented:
> I fixed the settings for linux and added some comment regarding macos via [8baf18b](https://github.com/holochain/holochain-infra/commit/8baf18b25d76ac8762b2c7459166677fbe8436a6)

nice. could you push that commit here so the PR can be merged? or shall i close it?

steveej commented:
i applied the patch and pushed it. thanks!

DavHau commented:
I cannot push to holochain-infra, last time I checked. Cherry-picking is fine :)

DavHau commented on modules/nixos/shared.nix:
```diff
@@ -29,10 +29,20 @@
+    })
+    // lib.optionalAttrs pkgs.stdenv.isDarwin {
+      # run at all minutes that are a multiple of 7
+      intervals =
+        builtins.map (Minute: {inherit Minute;}) (builtins.genList (n: n * 7) 9);
```
```suggestion
      intervals = let
        freq = 9;
      in
        builtins.map (Minute: {inherit Minute;}) (builtins.genList (n: (60 / freq) * (n + 1)) freq);
```

DavHau replied:
Personally I'd prefer to just write the list elements manually. It's much easier in the readers eyes and still not too hard to modify as long as the list doesn't get too long

steveej replied:
the frequency i was approaching was 7 per hour, it just isn't evenly distributed. not convinced it's worth tweaking the formula.

```
nix-repl> (builtins.genList (n: n * 7) 9)
[ 0 7 14 21 28 35 42 49 56 ]

nix-repl> (builtins.genList (n: (60 / 9) * (n + 1)) 9)
[ 6 12 18 24 30 36 42 48 54 ]
```

if i put in 7 as the frequency in the suggested formula it's not evenly distributed either.

```
nix-repl> (builtins.genList (n: (60 / 7) * (n + 1)) 7)
[ 8 16 24 32 40 48 56 ]
```

what's more important is that we get [the upstream PR merged](https://github.com/LnL7/nix-darwin/pull/611) that implements multiple interval support.

DavHau replied:
> what's more important is that we get [the upstream PR merged](https://github.com/LnL7/nix-darwin/pull/611) that implements multiple interval support.

Until it's merged, we could just set:
```
launchd.daemons.nix-gc.serviceConfig.StartCalendarInterval = mkForce ...;
```

steveej replied:
do you think it's an issue to use my fork in the meantime? (btw, i gave you write permissions to it)

DavHau replied:
Sure we can use your fork as well.

DavHau commented on modules/nixos/shared.nix:
```diff
@@ -29,10 +29,20 @@
+      # run at all minutes that are a multiple of 7
+      # verified with `systemd-analyze calendar "*-*-* *:0/7:00"`
+      dates = "*-*-* *:0/7:00";
+      # TODO: for some reason the timer is configured with RandomizedDelaySec=1800 regardless of this value
```
fixed in https://github.com/numtide/srvos/pull/84

steveej replied:
thanks! updated and rebased
